File: cs/algorithms_in/merge_sort.cs
namespace algorithms_in;

public class MergeSort
{
    public static void mergeSort(int[] array) {
        if (array.Length > 1) {
            int midpoint = array.Length / 2;
            int[] firstHalf = array[0..midpoint];
            mergeSort(firstHalf);
            
            int[] secondHalf = array[midpoint..array.Length];
            mergeSort(secondHalf);

            int[] temp = merge(firstHalf, secondHalf);
            
            for (int i = 0; i < temp.Length; i++) {
                array[i] = temp[i];
            }
        }
    }

    private static int[] merge(int[] firstHalf, int[] secondHalf) {
        int[] temp = new int[firstHalf.Length + secondHalf.Length];

        int current1 = 0;
        int current2 = 0;
        int current3 = 0;

        while (current1 < firstHalf.Length && current2 < secondHalf.Length) {
            if (firstHalf[current1] < secondHalf[current2]) {
                temp[current3++] = firstHalf[current1++];
            } else {
                temp[current3++] = secondHalf[current2++];
            }
        }

        while (current1 < firstHalf.Length) {
            temp[current3++] = firstHalf[current1++];
        }

        while (current2 < secondHalf.Length) {
            temp[current3++] = secondHalf[current2++];
        }

        return temp;
    }
}

File: go/merge_sort/merge_sort.go
package algorithms_in

func MergeSort(array []int) {
	if len(array) > 1 {
		var midpoint int = len(array) / 2
		var firstHalf []int = array[:midpoint]
		MergeSort(firstHalf)

		var secondHalf []int = array[midpoint:]
		MergeSort(secondHalf)

		var temp []int = merge(firstHalf, secondHalf)
		for i := 0; i < len(temp); i++ {
			array[i] = temp[i]
		}
	}
}

func merge(firstHalf []int, secondHalf []int) []int {
	var temp = make([]int, len(firstHalf)+len(secondHalf))

	var current1 int = 0
	var current2 int = 0
	var current3 int = 0

	for current1 < len(firstHalf) && current2 < len(secondHalf) {
		if firstHalf[current1] < secondHalf[current2] {
			temp[current3] = firstHalf[current1]
			current3++
			current1++
		} else {
			temp[current3] = secondHalf[current2]
			current3++
			current2++
		}
	}

	for current1 < len(firstHalf) {
		temp[current3] = firstHalf[current1]
		current3++
		current1++
	}

	for current2 < len(secondHalf) {
		temp[current3] = secondHalf[current2]
		current3++
		current2++
	}

	return temp
}